MdeModulePkg/Library/SmmCorePerformanceLib/MmCorePerformanceLib.c

```diff
@@ -0,0 +1,1374 @@
+      }
+
+      SmmCommData->BootRecordSize = mBootRecordSize;
+      break;
+
+    case SMM_FPDT_FUNCTION_GET_BOOT_RECORD_DATA:
+      DEBUG ((DEBUG_VERBOSE, "[%a] - SMM_FPDT_FUNCTION_GET_BOOT_RECORD_DATA entry.\n", __func__));
+      Status = EFI_UNSUPPORTED;
+      break;
+
+    case SMM_FPDT_FUNCTION_GET_BOOT_RECORD_DATA_BY_OFFSET:
+      DEBUG ((DEBUG_VERBOSE, "[%a] - SMM_FPDT_FUNCTION_GET_BOOT_RECORD_DATA_BY_OFFSET entry.\n", __func__));
+      BootRecordOffset = SmmCommData->BootRecordOffset;
+      BootRecordData   = SmmCommData->BootRecordData;
+      BootRecordSize   = SmmCommData->BootRecordSize;
+
+      if (BootRecordOffset >= mBootRecordSize) {
+        Status = EFI_INVALID_PARAMETER;
+        break;
+      }
+
+      if (BootRecordData == NULL) {
+        BootRecordData = (UINT8 *)SmmCommData + sizeof (SMM_BOOT_RECORD_COMMUNICATE);
+      } else if (!IsBufferOutsideMmValidInternal ((UINTN)BootRecordData, BootRecordSize)) {
+        Status = EFI_ACCESS_DENIED;
+        break;
+      }
+
+      //
+      // Sanity check
+      //
+      if (BootRecordSize > mBootRecordSize - BootRecordOffset) {
+        BootRecordSize = mBootRecordSize - BootRecordOffset;
+      }
+
+      // Note: Comm size passed to this handler already has OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data) removed.
+      if ((SmmCommData->BootRecordData == NULL) && (BootRecordSize - BootRecordOffset > TempCommBufferSize - sizeof (SMM_BOOT_RECORD_COMMUNICATE))) {
+        Status = EFI_BUFFER_TOO_SMALL;
+        break;
+      }
+
+      SmmCommData->BootRecordSize = BootRecordSize;
+      BootRecordBuffer            = ((UINT8 *)(mSmmBootPerformanceTable)) + sizeof (SMM_BOOT_PERFORMANCE_TABLE);
+      CopyMem (
+        (UINT8 *)BootRecordData,
+        BootRecordBuffer + BootRecordOffset,
+        BootRecordSize
+        );
+      mFpdtDataIsReported = TRUE;
+      break;
+
+    default:
+      Status = EFI_UNSUPPORTED;
+  }
+
+  SmmCommData->ReturnStatus = Status;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  This is the Event call back function is triggered in SMM to notify the Library
+  the system is entering runtime phase.
+
+  @param[in] Protocol   Points to the protocol's unique identifier
+  @param[in] Interface  Points to the interface instance
+  @param[in] Handle     The handle on which the interface was installed
+
+  @retval EFI_SUCCESS SmmAtRuntimeCallBack runs successfully
+ **/
+EFI_STATUS
+EFIAPI
+SmmCorePerformanceLibExitBootServicesCallback (
+  IN CONST EFI_GUID  *Protocol,
+  IN VOID            *Interface,
+  IN EFI_HANDLE      Handle
+  )
+{
+  //
+  // Disable performance measurement after ExitBootServices because
+  // 1. Performance measurement might impact SMI latency at runtime;
+  // 2. Performance log is copied to non SMRAM at ReadyToBoot so runtime performance
+  //    log is not useful.
+  //
+  mPerformanceMeasurementEnabled = FALSE;
+
+  return EFI_SUCCESS;
+}
+
+VOID
+InitializeMmCorePerformanceLibCommon (

```

Comment:
Missing function description?

Reply:
Also, i think it might be better to propagate the return code all the way to the constructor?

Reply:
Updated to propagate the error.